FEAT: initial correlation setup (#92)
arcus-azure/arcus.webapi: docs/features/correlation.md

---
title: "Correlate between HTTP requests/responses via ASP.NET Core middleware"
layout: default
---

# Correlation Between HTTP Requests

The `Arcus.WebApi.Correlation` library provides a way to add correlation between HTTP requests. 

> This correlation happens with the use of two HTTP request/response headers: `RequestId` and `X-Transaction-ID`.

Reviewer: ```suggestion

This correlation is based the `RequestId` and `X-Transaction-ID` HTTP request/response headers, however, you can fully configure different headers in case you need to.
```

Author: Yes 🥇 , that's more clear.

## How This Works

When an application is configured to use the default configuration of the correlation, each HTTP response will get an extra header called `RequestId` containing an unique identifier to distinguish between requests/responses.

> The `X-Transaction-ID` can be overriden by the request, meaning: if the HTTP request already contains a `X-Transaction-ID` header, the same header+value will be used in the HTTP response.

Reviewer: ```suggestion

The `X-Transaction-ID` can be overriden by the request, meaning: if the HTTP request already contains a `X-Transaction-ID` header, the same header+value will be used in the HTTP response.
```

Additional [configuration](#configuration) is available to tweak this functionality.

## Installation

> This feature requires to install our NuGet package

Reviewer: ```suggestion
This feature requires to install our NuGet package:
```

```shell
PM > Install-Package Arcus.WebApi.Correlation
```

## Usage

To make sure the correlation is added to the HTTP response, following additions have to be made in the `.ConfigureServices` and `Configure` methods:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddCorrelation();
}

public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.UserCorrelation();

    app.UseMvc();
}
```

Note: because the correlation is based on <span>ASP.NET</span> Core middleware, it's recommended to place it before the `.UseMvc` call.

## Configuration

Some extra options are available to alter the functionality of the correlation:

```csharp

> public void ConfigureServices(IServiceCollection services)

Reviewer: This is great, but we should mention what the defaults are

Author: Yeah, hmm. all the values set here are the defaults; but I guess indeed that's not clear.
Maybe a table of some sort with type+default?

Reviewer: You can also keep the current approach and just add a `(default: true)`?

If we would switch to a table, I would remove the docs below then and use table instead.

Up to you to decide, both are ok.

Author: Ah, yes. Thanks. That would be better.

Reviewer: Nitpick - Maybe move it at the end instead of beginning?
Is a bit weird to read.

Author: No problem, will do.
